spec/models/spree/product_sort_by_stocks_spec.rb

```diff
@@ -0,0 +1,93 @@
+# frozen_string_literal: true
+
+require 'spec_helper'
+
+RSpec.describe 'ProductSortByStocks' do
+  let(:product) { create(:product) }
+
+  describe 'class level SQL accessors' do
+    it 'exposes SQL Arel nodes for sorting' do
+      expect(Spree::Product.on_hand_sql).to be_a(Arel::Nodes::SqlLiteral)
+      expect(Spree::Product.on_hand_sql.to_s).to include('SELECT COALESCE(SUM')
+
+      expect(Spree::Product.backorderable_priority_sql).to be_a(Arel::Nodes::SqlLiteral)
+      expect(Spree::Product.backorderable_priority_sql.to_s).to include('CASE WHEN')
+    end
+  end
+
+  describe 'on_hand ransacker behaviour' do
+    it 'can be sorted via ransack by on_hand' do
+      product1 = create(:product)
+      product2 = create(:product)
+
+      product1.variants.first.stock_items.update_all(count_on_hand: 2)
+      product2.variants.first.stock_items.update_all(count_on_hand: 7)
+
+      # ransack sort key: 'on_hand asc' should put product1 before product2
+      result = Spree::Product.ransack(s: 'on_hand asc').result.to_a
+
+      expect(result.index(product1)).to be < result.index(product2)
+    end
+  end
+
+  describe 'backorderable_priority ransacker behaviour' do
+    it 'can be sorted via ransack by backorderable_priority' do
+      product1 = create(:product)
+      product2 = create(:product)
+
+      [product1, product2].each { |p|
+        p.variants.each { |v|
+          v.stock_items.update_all(backorderable: false)
+        }
+      }
+      product2.variants.first.stock_items.update_all(backorderable: true)
+
+      result = Spree::Product.ransack(s: 'backorderable_priority desc').result.to_a
+
+      expect(result.index(product2)).to be < result.index(product1)
+    end
+  end
+
+  describe 'combined sorting' do
+    # shared products for combined sorting examples
+    let!(:low) { create(:product) }
+    let!(:mid) { create(:product) }
+    let!(:high) { create(:product) }
+
+    it 'supports combined sorting: backorderable_priority (on-demad) then on_hand (asc)' do
+      low.variants.first.stock_items.update_all(count_on_hand: 1)
+      mid.variants.first.stock_items.update_all(count_on_hand: 5)
+      high.variants.first.stock_items.update_all(count_on_hand: 10)
+
+      # But make 'mid' backorderable so it should be sorted before 'low' when backorderable_priority asc
+      mid.variants.first.stock_items.update_all(backorderable: true)
+
+      # Controller transforms 'on_hand asc' into ['backorderable_priority asc', 'on_hand asc']
+      result = Spree::Product.ransack(s: ['backorderable_priority asc',
+                                          'on_hand asc']).result.to_a
+
+      # backorderable_priority asc places non-backorderable products first,
+      # so low (non-backorderable) should appear before mid (backorderable)
+      expect(result.index(low)).to be < result.index(mid)
+      # and low should appear before high
+      expect(result.index(low)).to be < result.index(high)
```

Comment:
The comments make it understandable what you are doing but I think it would more readable to just list the products in expected order, something like  : 
```suggestion
      expect(result[0]).to be low
      expect(result[2]).to be high
      expect(result[1]).to be mid
```